tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala

```diff
@@ -0,0 +1,249 @@
+package at.forsyte.apalache.io.itf
+
+import at.forsyte.apalache.io.itf.ItfToTla._
+import at.forsyte.apalache.io.json.{JsonDeserializationError, JsonRepresentation, ScalaFactory}
+import at.forsyte.apalache.io.lir.Trace
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.types.parser.DefaultType1Parser
+import at.forsyte.apalache.tla.types.{tla, ModelValueHandler}
+
+/**
+ * Reads a sequence of States, i.e. mappings from variable names to TlaEx values, from an ITF json.
+ *
+ * @tparam T
+ *   Any class extending JsonRepresentation
+ * @author
+ *   Jure Kukovec
+ */
+class ItfToTla[T <: JsonRepresentation](
+    scalaFactory: ScalaFactory[T]) {
+
+  /**
+   * Checks that the Apalache-specific ITF requirements are met, as defined in ADR15. If yes, returns a mapping from
+   * variables to their types.
+   */
+  private[itf] def validateShapeAndGetTypes(json: T): Map[String, TlaType1] = {
```

Comment:
What about making this and similar methods in here return an `Either` that tracked the error rather than throwing exceptions? It would make it easier to reason about the possible error cases, and you could drop the `validate` or `attempt` prefixes in the methods, since it'd be clear from the types that they could fail, and you can sequence with for comprehensions.

Reply:
This feels reminiscent of https://github.com/informalsystems/apalache/pull/2234